packages/lib/sync/ISyncService.ts

```diff
@@ -0,0 +1,77 @@
+import { DeploymentType } from "@prisma/admin-client";
+import { User } from "@prisma/client";
+
+import logger from "@calcom/lib/logger";
+import { default as webPrisma } from "@calcom/prisma";
+
+export type UserInfo = {
+  email: string;
+  name: string | null;
+  id: number;
+  username: string | null;
+};
+
+export type WebUserInfoType = UserInfo & {
+  plan: User["plan"];
+};
+
+export type ConsoleUserInfoType = UserInfo & {
+  plan: DeploymentType;
+};
+
+export interface IUserCreation<T> {
+  create(info: T): Promise<any>;
+  update(info: T): Promise<any>;
+  upsert?: never;
+}
+
+export interface IUserUpsertion<T> {
+  create?: never;
+  update?: never;
+  upsert(info: T): Promise<any>;
+}
+
+export interface ISyncService {
+  web: {
+    user: IUserCreation<WebUserInfoType> | IUserUpsertion<WebUserInfoType>;
+  };
+  console: {
+    user: IUserCreation<ConsoleUserInfoType> | IUserUpsertion<ConsoleUserInfoType>;
+  };
+}
+
+export default class SyncServiceCore {
```

Comment:
Core class to have Sync Services base on, mainly to concentrate common implementation among them all.